rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,230 @@
+
+The suggestion [here](https://groups.google.com/g/kubernetes-sig-architecture/c/mVGobfD4TpY/m/nkdbkX1iBwAJ) is to use the UID of the `kube-system` namespace, as this [cannot be deleted easily](https://groups.google.com/g/kubernetes-sig-architecture/c/mVGobfD4TpY/m/lR4SnaqpAAAJ) - so is practically speaking indelible.
+
+We can take the UID of that namespace and apply a hashing algorithm to result in 6-7 character code which can be used as the clusterID for the local cluster.
+
+#### Update GUID Logic
+
+Currently the DNS Operator builds the LB CNAME with a guid based on the name of the gateway, as this could potentially be different on different clusters, this will need to be updated to generate based on the zone's root hostname, to ensure it is consistently generated on all clusters that interact with this zone.
+
+#### Build DNS Records from the provider zone
+
+When the DNS Operator reconciles a kuadrantRecord and will first pull the most recent records from the DNS Provider and construct a separate zoneRecord.
+
+These zoneRecords will be created with the DNS Operator set as the owner, so they can be identified in the future. When reconciling a DNS Record is being reconciled by the DNS Operator, first check if it is owned by the DNS Operator and if so, do nothing - these DNS Records should never be reconciled.
+
+As this could result in many requests to the DNS Provider API a mechanism is required to reduce this to only request when required:
+
+#### Testing if a DNS Provider pull is Required
+
+When the DNS Operator publishes any changes to a zone, it will also add/update a (`kuadrant.last_modified_at`) TXT record with the current timestamp.
+
+If the DNS Operator has no local zoneRecords for a zone, it will list and build the zoneRecords.
+
+If the DNS Operator already has zoneRecords (and their lastModified time is older than the refreshInterval, more on this later) it will first execute a GET request for the `kuadrant.last_modified_at` TXT record, if the value of that field is more recent than the zoneRecords lastModified time, then the DNS Operator will list and rebuild it's zoneRecords.
+
+If the zoneRecords lastModified time is less old than the refreshInterval, a GET and possibly a LIST should only be executed when the spec of the kuadrantRecord has changed (as this will likely result in a write to the API).
+
+#### Ensure local records are accurate
+
+Once the zoneRecord is created / updated on cluster, the operator will then scan through this zoneRecord and remove any records for the local cluster. Then inject the values from the kuadrantRecord into the zoneRecord. After this the DNS Operator will sanity check the result and prune any bad records (we will expand on that later). Once the zoneRecord is updated and sanity checked, if it has changed then the result is sent back to the DNS Provider so that the zone can be updated and made accurate (with an added update to the `kuadrant.last_modified_at`) TXT record.
```

Comment:
I guess this is only needed if the observedGeneration of the kuadrantRecord is updated? IE if the spec is changed?

Reply:
````kuadrant.last_modified_at`) TXT record.```
This field can't be locked. So it is possible / likely that sometimes one controller will set this to an older value than another. Not sure if that matters. But this is a little bit of a fire and forget operation. I imagine the variation in time will be small however as the only thing that is triggering this is a change that needs to be written to the zone. So the value should be calculated immediately before the API call to make it as accurate as possible?

Reply:
no, we also need it if no changes have occurred locally to ensure we have all the health checks needed for other clusters changes.